### ISSUE TITLE
ValueError: Inferred frequency None from passed values does not conform to passed frequency C

### ISSUE THREAD

**Issue by Boubside** [quantopian/alphalens#367]
## ValueError: Inferred frequency None from passed values does not conform to passed frequency C

Hi,
I'm trying to create my full tear sheet using alphalens on my own environment, but when I run alphalens.utils.get_clean_factor_and_forward_returns() I get the following error :

**Full Traceback:**
```python
Traceback (most recent call last):
  File "main.py", line 23, in <module>
    periods=(1,5,10))
  File "/home/sj/.local/lib/python3.6/site-packages/alphalens/utils.py", line 793, in get_clean_factor_and_forward_returns
    cumulative_returns)
  File "/home/sj/.local/lib/python3.6/site-packages/alphalens/utils.py", line 341, in compute_forward_returns
    df.index.levels[0].freq = freq
  File "/home/sj/.local/lib/python3.6/site-packages/pandas/core/indexes/datetimelike.py", line 241, in freq
    self._validate_frequency(self, value)
  File "/home/sj/.local/lib/python3.6/site-packages/pandas/core/indexes/datetimelike.py", line 230, in _validate_frequency
    raise ValueError(msg.format(infer=inferred, passed=freq.freqstr))
ValueError: Inferred frequency None from passed values does not conform to passed frequency C

```

**Code :**
```pyrhon
factor_data = alphalens.utils.get_clean_factor_and_forward_returns(signal,
                                                                   pricings,
                                                                   quantiles=2,
                                                                   periods=(1,5,10))

alphalens.tears.create_full_tear_sheet(factor_data)
```

where signal is :
```
date                asset          
...

2019-11-01 09:56:00 A      0.000000
                    AAL    0.000000
2019-11-01 09:57:00 A      0.000000
                    AAL    0.000000
2019-11-01 09:58:00 A      0.000000
                    AAL   -0.142857
2019-11-01 09:59:00 A      0.000000
                    AAL    0.000000
2019-11-01 10:00:00 A      0.000000
                    AAL   -1.264151

...
```
and pricing :
```
                          A     AAL
Date                               
...

2019-11-01 09:36:00  76.550  30.390
2019-11-01 09:37:00  76.550  30.405
2019-11-01 09:38:00  76.655  30.400
2019-11-01 09:39:00  76.655  30.375
2019-11-01 09:40:00  76.655  30.285

...
```

I made sure to have enough data in pricings compared to my factor. 
I'm not sure if the is a bug or a misunderstanding from my side.

## Versions

* Alphalens version: 0.3.6
* Python version: 3.6.9
* Pandas version: 0.23.4
* Matplotlib version: 3.1.2
* Numpy version: 1.18.1
* Linux ubuntu 18.04

**Comment by luca-s** [comment]
Hi @Boubside. Does `pricing` contain a value for each datetime in your `signal` index?

**Comment by luca-s** [comment]
@Boubside I also have some doubts regarding `signal`. It contains only 2 assets but you pass `quantiles=2`. What exactly do you expect as output from Alphalens? You might get better results with `bins` option insted of `quantiles`. Even though I believe you might be interested more on a Event Study tearsheet maybe (you can find [here](https://github.com/quantopian/alphalens/blob/master/alphalens/examples/alphalens_tutorial_on_quantopian.ipynb) an explanation?

**Comment by Boubside** [comment]
Hi @luca-s. When loading my prices, I forward fill NAN values and drop any remaining NAN for each single asset before stacking. I checked my indexes using the following code :
```
signal_unstack = signal.unstack()
pricings.dropna(inplace=True)
signal_unstack.dropna(inplace=True)
pricings_cropped = pricings.loc[signal_unstack.index.min():signal_unstack.index.max()]
print(pricings_cropped.index.difference(signal_unstack.index))
```
Dropping NANs once again to make sure no value is missing and got an empty DateTimeIndex output so I think pricing have value for every index of signal, plus a lot of extra both at the beginning and at the end. I also tried that on one trading day and it looks like I have no voids in my pricing data :
```
print(pd.date_range(pricings_cropped.index.min(),pricings_cropped.index.max(), freq="1min").difference(pricings.index))
```

My signal is computed from the pricing so I don't think I have anything more in signal than in pricings.



To answer your second comment, its in fact my first time using alphalens, and this was just supposed to be a test before starting some 'real' analysis. By reading the tutorial you link I understand that quantiles create buckets by assets and bins create buckets by individual values right ? My goal was to analyse the predictive power of my signal, which by the way is a dummy signal just to test alphalens once again.
I tested using more assets, with both bins and quantiles defined as an int or array of int and get the same error.

**Comment by Boubside** [comment]
@luca-s, by reading a little more of the tutorial, the periods argument is mentioned as being in days, is it because tutorial data is in days or do I need to define periods differently to have them in minutes ? I'm using minutes data and want the periods to be under 10 minutes.

**Comment by luca-s** [comment]
> By reading the tutorial you link I understand that quantiles create buckets by assets and bins create buckets by individual values right ?

Not exactly. Both are based on the values. `quantiles` option chooses the value range in each bucket so that each bucket contains the same number of assets (same number of elements per bucker is  costant, the value range is variable).  `bins` option chooses the value range to be identical for each bucker (number of elements per bucket is variable, value range is constant).

>  by reading a little more of the tutorial, the periods argument is mentioned as being in days, is it because tutorial data is in days or do I need to define periods differently to have them in minutes ? I'm using minutes data and want the periods to be under 10 minutes.

That was a simplification of the tutorial that covers the most common scenario of Alphalens usage on Quantopian platform that is daily factors. Your use case should work fine since periods don't represent neither days nor anything else but are data dependent. I'll copy paste here the description of `pricing` from `utils.get_clean_factor_and_forward_returns`

```
prices : pd.DataFrame
        A wide form Pandas DataFrame indexed by timestamp with assets
        in the columns.
        Pricing data must span the factor analysis time period plus an
        additional buffer window that is greater than the maximum number
        of expected periods in the forward returns calculations.
        It is important to pass the correct pricing data in depending on
        what time of period your signal was generated so to avoid lookahead
        bias, or  delayed calculations.
        'Prices' must contain at least an entry for each timestamp/asset
        combination in 'factor'. This entry should reflect the buy price
        for the assets and usually it is the next available price after the
        factor is computed but it can also be a later price if the factor is
        meant to be traded later (e.g. if the factor is computed at market
        open but traded 1 hour after market open the price information should
        be 1 hour after market open).
        'Prices' must also contain entries for timestamps following each
        timestamp/asset combination in 'factor', as many more timestamps
        as the maximum value in 'periods'. The asset price after 'period'
        timestamps will be considered the sell price for that asset when
        computing 'period' forward returns.
```

Would you  be able to extract a minimum amount of data from your signal and price that reproduces the bug and copy/paste the minimum example here (e.g. define signal and pricing as two DataFrame variables with some hand writted data) so that I can debug the code and find out why you get the error?

**Comment by Boubside** [comment]
@luca-s , here a very basic example that raise the same error :
```
import pandas as pd
import datetime as dt
import alphalens

price = {'A': [76.930, 76.730, 76.850, 76.750, 76.750, 76.880, 76.925], 'AAL': [30.240, 30.210, 30.210, 30.200, 30.150, 30.245, 30.315]}
factor = {'A': [0,0,0,0], 'AAL': [0,0,-0.142857,0]}

pricings = pd.DataFrame.from_dict(price)
pricings['date'] = pd.date_range(dt.datetime(2019,11,28),dt.datetime(2019,11,28) + dt.timedelta(minutes=6), freq="1min")
pricings.set_index(['date'], inplace=True)

signal = pd.DataFrame.from_dict(factor)
signal['date'] = pd.date_range(dt.datetime(2019,11,28),dt.datetime(2019,11,28) + dt.timedelta(minutes=3), freq="1min")
signal.set_index(['date'], inplace=True)
signal = signal.stack()

print(pricings)
print(signal)


factor_data = alphalens.utils.get_clean_factor_and_forward_returns(signal,
                                                                   pricings,
                                                                   bins = 2,
                                                                   periods=(1,2))
```

And the full corresponding output:
```
                          A     AAL
date                               
2019-11-28 00:00:00  76.930  30.240
2019-11-28 00:01:00  76.730  30.210
2019-11-28 00:02:00  76.850  30.210
2019-11-28 00:03:00  76.750  30.200
2019-11-28 00:04:00  76.750  30.150
2019-11-28 00:05:00  76.880  30.245
2019-11-28 00:06:00  76.925  30.315
date                    
2019-11-28 00:00:00  A      0.000000
                     AAL    0.000000
2019-11-28 00:01:00  A      0.000000
                     AAL    0.000000
2019-11-28 00:02:00  A      0.000000
                     AAL   -0.142857
2019-11-28 00:03:00  A      0.000000
                     AAL    0.000000
dtype: float64
Traceback (most recent call last):
  File "main.py", line 48, in <module>
    periods=(1,2))
  File "/home/sj/.local/lib/python3.6/site-packages/alphalens/utils.py", line 793, in get_clean_factor_and_forward_returns
    cumulative_returns)
  File "/home/sj/.local/lib/python3.6/site-packages/alphalens/utils.py", line 341, in compute_forward_returns
    df.index.levels[0].freq = freq
  File "/home/sj/.local/lib/python3.6/site-packages/pandas/core/indexes/datetimelike.py", line 241, in freq
    self._validate_frequency(self, value)
  File "/home/sj/.local/lib/python3.6/site-packages/pandas/core/indexes/datetimelike.py", line 230, in _validate_frequency
    raise ValueError(msg.format(infer=inferred, passed=freq.freqstr))
ValueError: Inferred frequency T from passed values does not conform to passed frequency C
```


The problem might come from the fact that my signal is very sparse, but even with manually defined bins, I have the same error (using bins=[-10,0,10], maybe this is not the correct wai to do it...)

**Comment by luca-s** [comment]
It turned out to be a compatibility version with recent pandas releases. For Alphalens to work you need pandas==0.20.3. I would suggest to create a python environment with version 0.20.3 of pandas for Alphalens, until the module is updated to work with recent pandas versions.

**Comment by Boubside** [comment]
Ok, I suspected something like that but I used 0.23.4 because of alphalens v0.3.5 release note.
Thank you very much !

**Comment by qiuwei** [comment]
@luca-s  I thought it was already solved by https://github.com/quantopian/alphalens/pull/328, wasn't it?

**Comment by luca-s** [comment]
@qiuwei I thought it too, but since pandas 0.23.4 is not part of the automated tests it might be that some incompatible changes has been included in Alphalens code.

**Comment by mad-ada** [comment]
@Boubside I came cross the same issue today. When checking the source code, I find alphalens only support three  time series frequencies (**_CustomBusinessDay, Day, BusinessDay_**). But what I need is **_Minute_** freq.

A quick fix that works for me: 
1. copy out _alphalens.utils.py_, and change the 'freq' parameter in function _compute_forward_returns_. e.g. in my case, **df.index.levels[0].freq = '5T'**.
 **NOTE:** Need to make sure the passing in parameters: Both factor and prices are tz_aware and have the same frequency (these parameters can be easily set when creating the dataframe).
2. Furthermore, can try passing in **freq** as a parameter when generating clean factor and return data to support different frequency data. 

Hope this helps.

**Comment by longliveh** [comment]
price and signal datetime indexes need to be the same

**Comment by CasperChen0617** [comment]
find utils.py and the "Before" session code
'''
Before :     
  df.index.levels[0].freq = freq

After : 
  from pandas.tseries.offsets import Minute
  if isinstance(prices.index.freq, Minute):
      df.index.levels[0].freq = "T"
  else:
      df.index.levels[0].freq = freq
'''      
And you can fixed this problem
